### PR TITLE
Rdf ingestion mvp

### PR DESCRIPTION
**Pull request by stephengoldbaum**
## Summary

This PR introduces a new **RDF ingestion source** for DataHub, enabling ingestion of RDF/OWL ontologies (Turtle, RDF/XML, JSON-LD, N3, N-Triples) with a focus on business glossaries. The source extracts glossary terms, term hierarchies, and relationships from RDF files using standard vocabularies like SKOS, OWL, and RDFS.

### Core Features
- **RDF Ingestion Source** (`type: rdf`) - Native DataHub plugin for RDF/OWL ontologies
- **Multiple Format Support** - Turtle, RDF/XML, JSON-LD, N3, N-Triples
- **Flexible Source Loading** - Files, directories (with recursive option), URLs, and comma-separated file lists
- **Glossary Term Extraction** - Converts `skos:Concept` and `owl:Class` to DataHub GlossaryTerms
- **Glossary Node Hierarchy** - Auto-creates glossary nodes from IRI path hierarchies
- **Term Relationships** - Extracts `skos:broader` and `skos:narrower` relationships as `isRelatedTerms`
- **Stateful Ingestion** - Full support for stale entity removal via `stateful_ingestion` config
- **Platform Instance Support** - Configurable platform instances via `platform_instance` config

### Architecture
- **Modular Entity Processing** - Clean separation with extractors, converters, and MCP builders
- **Dependency-Based Processing** - Topological sort for correct entity processing order
- **Test Connection Support** - Implements `test_connection()` for connection validation

## Capabilities

The source supports the following DataHub capabilities:

| Capability              | Status | Notes                                                         |
| ----------------------- | :----: | ------------------------------------------------------------- |
| Glossary Terms          |   ✅   | Enabled by default                                            |
| Glossary Nodes          |   ✅   | Auto-created from IRI path hierarchies                        |
| Term Relationships      |   ✅   | Supports `skos:broader` and `skos:narrower`                   |
| Detect Deleted Entities |   ✅   | Requires `stateful_ingestion.enabled: true`                   |
| Platform Instance       |   ✅   | Supported via `platform_instance` config                      |
| Extract Descriptions    |   ✅   | Enabled by default (from `skos:definition` or `rdfs:comment`) |
| Data Domain             |   ❌   | Not applicable (domains used internally for hierarchy)        |
| Dataset Profiling       |   ❌   | Not applicable                                                |
| Extract Lineage         |   ❌   | Not in MVP                                                    |
| Extract Ownership       |   ❌   | Not supported                                                 |
| Extract Tags            |   ❌   | Not supported                                                 |

### Test Coverage
- **126 unit tests** - Comprehensive coverage of core functionality, error handling, and edge cases
- **16 integration tests** - End-to-end testing with golden file validation
- **Test scenarios include**:
  - Simple glossary ingestion
  - Glossary with relationships
  - Glossary with domains
  - Multiple RDF formats (Turtle, RDF/XML, JSON-LD)
  - Recursive directory ingestion
  - Export filtering (`export_only`, `skip_export`)
  - Stateful ingestion with stale entity removal
  - Error handling (missing files, malformed RDF, invalid formats)
  - Large file performance warnings
  - Path traversal protection

### Test Files
- `tests/unit/rdf/` - Unit tests for individual components
- `tests/integration/rdf/` - Integration tests with golden file validation
- All tests passing ✅

### User Documentation
- **`docs/sources/rdf/rdf.md`** - Comprehensive user guide (489 lines)
  - Quickstart guide
  - Configuration reference
  - RDF format and source types
  - Dialects and selective export
  - Stateful ingestion guide
  - Example RDF files
  - IRI-to-URN mapping
  - Glossary node hierarchy
  - Supported vocabularies
  - Limitations and troubleshooting

### Recipe Examples
- **`docs/sources/rdf/rdf_recipe.yml`** - Example recipes for basic and stateful ingestion

### Integration Test Documentation
- **`tests/integration/rdf/README.md`** - Detailed guide for running integration tests

## Configuration Example

source:
  type: rdf
  config:
    source: ./glossary.ttl
    format: turtle
    environment: PROD
    stateful_ingestion:
      enabled: true
      remove_stale_metadata: true
    export_only:
      - glossary## Files Changed

#### Security & Performance
- **URL Loading Security** - Timeout limits, size limits, and redirect limits for safe URL loading
- **Path Traversal Protection** - Configurable enforcement to prevent access outside intended directories
- **Memory Efficiency** - Generator patterns for work unit generation and streaming for large URL downloads
- **Format Validation** - Validates RDF formats before processing

#### Code Quality
- **Thread-Safe Registry** - Entity registry uses double-checked locking pattern for thread safety
- **Component Validation** - Validates registered components for entity type consistency
- **Type Safety** - Complete type hints with proper forward references for MCP return types
- **Error Handling** - Granular error reporting with structured logs and context
- **URN Generation** - Standardized URN format using dot notation, proper encoding, and GUID fallback for non-ASCII characters

### New Files
- `src/datahub/ingestion/source/rdf/ingestion/rdf_source.py` - Main source implementation
- `src/datahub/ingestion/source/rdf/core/rdf_loader.py` - RDF loading utilities with security
- `src/datahub/ingestion/source/rdf/core/urn_generator.py` - URN generation with encoding
- `src/datahub/ingestion/source/rdf/entities/base.py` - Base interfaces for entity processing
- `src/datahub/ingestion/source/rdf/entities/registry.py` - Thread-safe entity registry
- `docs/sources/rdf/rdf.md` - User documentation
- `docs/sources/rdf/rdf_recipe.yml` - Recipe examples
- `tests/integration/rdf/test_rdf_source.py` - Integration tests
- `tests/unit/rdf/` - Unit tests (multiple files)

### Modified Files
- `setup.py` - Added RDF source to entry points (line 862)

## Breaking Changes

**None** - This is a new feature addition with no breaking changes to existing functionality.

## Support Status

The RDF source is marked as **INCUBATING** (`SupportStatus.INCUBATING`), indicating it's ready for community adoption but may have minor version changes in future releases based on feedback.

## Checklist

- [x] Plugin registered in `setup.py`
- [x] Source class properly decorated (`@platform_name`, `@config_class`, `@support_status`)
- [x] Capability decorators added
- [x] Stateful ingestion implemented
- [x] `test_connection()` implemented
- [x] Comprehensive error handling
- [x] Security measures (timeouts, size limits, path traversal protection)
- [x] Memory-efficient patterns (generators, streaming)
- [x] Thread-safe registry
- [x] Type hints complete
- [x] All tests passing
- [x] User documentation complete
- [x] Integration test documentation
- [x] Code follows DataHub standards
- [x] Linting passes

### CODE REVIEW

File: autogenerated/capability_summary.json [hunk 1]
```diff
@@ -0,0 +1,375 @@
+{
```
Comment (gabe-lyons) [inline autogenerated/capability_summary.json]: why is this PR creating this new file?

Comment (stephengoldbaum) [reply autogenerated/capability_summary.json]: It was updated to add the RDF capabilities and then regenerated

**Comment by rajatoss** [comment]
## Connector Tests Results

**All connector tests passed** for commit `cecd1b8`

[View full test logs →](https://github.com/acryldata/connector-tests/actions/runs/22786971562)

*Autogenerated by the connector-tests CI pipeline.*

File: autogenerated/capability_summary.json (hunk 1 above)
Comment (gabe-lyons) [reply autogenerated/capability_summary.json]: this doesn't feel like a sufficient answer, because why would the file be net new in that case? regenerated would imply a diff. this file is net new.

File: datahub-web-react/src/app/ingestV2/source/builder/RecipeForm/rdf.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
+import { FieldType, RecipeField } from '@app/ingestV2/source/builder/RecipeForm/common';
+
+export const RDF_SOURCE: RecipeField = {
+    name: 'source',
+    label: 'Source',
+    helper: 'RDF source: file, folder, URL, or comma-separated files',
+    tooltip:
+        'Source to process: file path, folder path, server URL, or comma-separated files. Examples: /path/to/file.ttl, /path/to/folder, https://example.com/data.ttl, file1.ttl,file2.ttl',
```
Comment (gabe-lyons) [inline datahub-web-react/src/app/ingestV2/source/builder/RecipeForm/rdf.ts]: @stephengoldbaum if this is only a path to local files, there is no purpose in having RDF in the UI. there is no capacity for users to upload files to datahub. this needs to support remote files, or will be nearly useless.

File: autogenerated/capability_summary.json (hunk 1 above)
Comment (stephengoldbaum) [reply autogenerated/capability_summary.json]: Ack. File removed.

File: datahub-web-react/src/app/ingestV2/source/builder/RecipeForm/rdf.ts (hunk 2 above)
Comment (stephengoldbaum) [reply datahub-web-react/src/app/ingestV2/source/builder/RecipeForm/rdf.ts]: I've updated to reflect the preferred approach is URL and local file only works for CLI.
